mitiq/shadows/post_processing.py

```diff
@@ -0,0 +1,141 @@
+
+    # local unitary that applied to the computational basis states, e.g.
+    # $$Z-$$basis measurement,
+    # which equivalent to a random Pauli measurement, i.e. for each qubit,
+    # we randomly decide to measure the Pauli operators.
+    phase_z = np.array([[1, 0], [0, -1j]], dtype=np.complex128)
+    hadamard = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
+    identity = np.eye(2)
+
+    # act $$Ad_{U^\dagger}$$ on the computational basis states, where $$U$$ is
+    # Define ensemble which the random unitary sample from.
+    unitaries = [hadamard, hadamard @ phase_z, identity]
+
+    # reconstructing a single snapshot state by applying Eq. (S44)
+    rho_snapshot = np.array([1], dtype=np.complex128)
+    for i in range(num_qubits):
+        state = zero_state if b_list[i] == 1 else one_state
+        U = unitaries[PAULI_MAP[u_list[i]]]
+
+        # applying Eq. (S44)
+        local_rho = 3 * (U.conj().T @ state @ U) - identity
+        rho_snapshot = np.kron(rho_snapshot, local_rho)
+
+    return rho_snapshot
+
+
+def shadow_state_reconstruction(
+    measurement_outcomes: Tuple[np.ndarray[Any, Any], np.ndarray[Any, Any]]
+) -> np.ndarray[Any, Any]:
+    """
+    Reconstruct a state approximation as an average over all snapshots.
+
+    Args:
+        measurement_outcomes (tuple): A shadow tuple obtained
+        from `z_basis_measurement`.
+
+    Returns:
+        Numpy array with the reconstructed quantum state.
+    """
+    num_snapshots, num_qubits = measurement_outcomes[0].shape
+
+    # classical values
+    b_lists, u_lists = measurement_outcomes
+
+    # Averaging over snapshot states.
+    shadow_rho = np.zeros(
+        (2**num_qubits, 2**num_qubits), dtype=np.complex128
+    )
+    for i in range(num_snapshots):
+        shadow_rho += snapshot_state(b_lists[i], u_lists[i])
+
+    return shadow_rho / num_snapshots
+
+
+def expectation_estimation_shadow(
+    measurement_outcomes: Tuple[np.ndarray[Any, Any], np.ndarray[Any, Any]],
+    observable: Any,
+    k: int,
+) -> float:
+    # Need R = NK in total, and split into K subsets of size and
+    # N is the number of snapshots.
+    # each subset is a tuple of (b_lists, u_lists) and each
+    # element of the list is a list of length len(qubits)
+
+    r"""
+    Calculate the estimator $$E[O] = median(Tr{rho_{(k)} O})$$ where
+    $$rho_(k))$$is set of $$k$$ snapshots in the shadow.
+    Use median of means to ameliorate the effects of outliers.
+
+    Args:
+        measurement_outcomes (tuple): A shadow tuple obtained from
+        `shadow_measure_with_executor`.
+        observable (cirq.PauliString): Single cirq observable consisting of
+        single Pauli operators e.g. cirq.X(0) * cirq.Y(1).
+        k (int): number of splits in the median of means estimator. k * N = R,
+        where R is the total number of measurements,
+         N is the number of snapshots.
+
+    Returns:
+        Scalar corresponding to the estimate of the observable.
+    """
```

Comment:
Make sure to check how the docs look [[link](https://mitiq--1908.org.readthedocs.build/en/1908/apidoc.html#mitiq.shadows.post_processing.snapshot_state)] when building them. Right now it needs some work to be formatted correctly.

Reply:
Thanks! I've resolved this issue in previous commits.